File: api/sections.js
const express = require('express');
const axios = require('axios');
const app = express();
const http = require('http');
const https = require('https');
// First port on V2 
const port = 3019;
const API_KEY = 'cf23df2207d99a74fbe169e3eba035e633b65d94'; 
// const username = 'brd-customer-hl_e7275c4d-zone-dropchecker';
// const password = 'tbosd0xrywgv';
const username = 'brd-customer-hl_e7275c4d-zone-padelis_web_unlocker_main';
const password = 'tikqvzik1t9p'
const proxyPort = 22225;

const httpAgent = new http.Agent({ keepAlive: true });
const httpsAgent = new https.Agent({ keepAlive: true, rejectUnauthorized: false });

app.use(express.json());

const { Pool } = require('pg');

const pool = new Pool({
    host: 'database-1-instance-1.cwysyy38ldpw.us-east-2.rds.amazonaws.com',
    database: 'tryAgain',
    user: 'isdsar',
    password: 'Outeclanisdsar123*',
    port: 5432
});

async function getEndpointURL(event_code, link_id) {
  const query = `
    SELECT DISTINCT
        p.event_code,
        p.link_id,
        p.data_acc_id,
        p.distributor_id,
        CONCAT(
            'https://',
            p.venue_name,
            '.evenue.net/cgi-bin/ncommerce3/SEPyos?linkID=',
            p.link_id,
            '&get=sect&itC=',
            p.event_code,
            '&data_acc_id=',
            p.data_acc_id,
            '&link_id=',
            p.link_id,
            '&distributor_id=',
            p.distributor_id,
            '&eventcode_id=', 
            p.event_code
        ) AS sections_link_url
    FROM
        production_evenuemap p  
    INNER JOIN tmEvents t on p.event_code = t.event_code 
    WHERE 
        t.start_date IS NOT NULL 
        AND p.event_code = $1
        AND p.link_id = $2;
  `;

  const { rows } = await pool.query(query, [event_code, link_id]);
  return rows[0].sections_link_url;
}

async function makeRequest(data, res) {
  try {
    
    const parts = data.split('?');
    const url = parts[0] + '?' + parts[1];  // construct URL up to event code
    
    // const parts = data.split('+');
    // const url = parts[0];
    // const searchParams = new URLSearchParams(parts[1]); // treat remaining parts as the parameters
    console.log(url);
    // console.log(searchParams);
    const searchParams = new URLSearchParams(url.split('?')[1]); // treat remaining parts as the parameters
    const venue_name_id = url.match(/https:\/\/(.*?)\.evenue\.net/)[1]; 
    const link_id = searchParams.get('link_id');
    const data_acc_id = searchParams.get('data_acc_id');
    const distributor_id = searchParams.get('distributor_id');
    const eventcode_id = searchParams.get('eventcode_id');

    const session_id = (1000000 * Math.random()) | 0;
    // 198.13.33.79
    // const super_proxy_url = `http://${username}-session-${session_id}:${password}@198.13.33.79:${proxyPort}`;
    const super_proxy_url = `http://${username}-session-${session_id}:${password}@brd.superproxy.io:${proxyPort}`;
    const superProxyUrl = new URL(super_proxy_url);

    const options = {
      method: 'get',
      url: url,
      proxy: {
        host: superProxyUrl.hostname,
        port: superProxyUrl.port,
        auth: {
          username: superProxyUrl.username,
          password: superProxyUrl.password,
        },
      },
      httpAgent: httpAgent, // use the shared agent
      httpsAgent: httpsAgent, // use the shared agent
      headers: {
        'Pac-Context-Data': JSON.stringify({
          distributorId: distributor_id,
          dataAccountId: data_acc_id,
          siteId: "ev_"+link_id, 
        }),
        'Pac-Authz-Data': JSON.stringify({
            distributorId: distributor_id,
            dataAccountId: data_acc_id,
            iProfileId: ""
        }),
      },
    }

    let response = await axios(options);
    // console.log(response);
    // Handle the response
    if (response.status === 200) {
        response = response.data; 
        console.log(`Successful GETSECTIONS request for: ${eventcode_id} and ${link_id}\n`);
        const front_end_url = `http://${venue_name_id}.evenue.net/cgi-bin/ncommerce3/SEGetEventInfo?linkID=${link_id}&ticketCode=${eventcode_id}&groupCode=&shopperContext=&pc=&caller=&appCode=&cgc=&prc=&ppc=&psc=&sm=1&dataAccId=${data_acc_id}&siteId=ev_${link_id}`;    
        // add the additional properties to the response object
        response.event_code = eventcode_id; 
        response.data_acc_id = data_acc_id; 
        response.link_id = link_id; 
        response.venue_name = venue_name_id; 
        response.status = 200; 

        // check if the response has the specific status and message
        if (response.status.mg === 'Unable to get seat map information.  Please try again later.') {
          console.log('Seatmap blocked: '+front_end_url+'\n');
          res.write(`${JSON.stringify(null)}\n\n`);
          res.end();
        }
        else {      
          res.write(`${JSON.stringify(response)}\n\n`);
          res.end();        
        } 
      }

  } catch (error) {    
    // console.log(error);
    // Check if error.response exists, meaning the server responded with an error status code
    if (error.response) {
        // Extract status code and message from the error response
        const { status, statusText } = error.response;
        
        // Handle different HTTP status codes
        if (status === 400) {
            res.write(`${JSON.stringify({ error: 'Bad Request', message: statusText, 'status': status })}\n\n`);
        } else if (status === 403) {
            res.write(`${JSON.stringify({ error: 'Forbidden', message: statusText, 'status': status })}\n\n`);
        } else if (status === 500) {
            res.write(`${JSON.stringify({ error: 'Internal Server Error', message: statusText, 'status': status })}\n\n`);
        } else if (status === 404) {
            res.write(`${JSON.stringify({ error: 'Not Found', message: statusText, 'status': status })}\n\n`);
        } else {
            // Default case for other status codes
            res.write(`${JSON.stringify({ error: 'Error in making request', message: statusText, 'status': status })}\n\n`);
        }
    } else if (error.request) {
        // The request was made but no response was received
        res.write(`${JSON.stringify({ error: 'No response received from server', message: 'The server did not respond to the request' })}\n\n`);
    } else {
        // Something happened in setting up the request that triggered an Error
        res.write(`${JSON.stringify({ error: 'Request Setup Error', message: error.message })}\n\n`);
    }
    res.end();
    }
}

app.use((req, res, next) => {
  const apiKey = req.header('X-API-Key');
  if (!apiKey || apiKey !== API_KEY) res.status(401).json({ message: 'Not Authorized' });
  else next();
});

app.get('/production-isolatedsections', async (req, res) => {
  res.setHeader('Content-Type', 'application/json'); // application/json text/event-stream
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();

  try {
    const { event_code, link_id } = req.query;
    const data = await getEndpointURL(event_code, link_id);
    await makeRequest(data, res);

  } catch (err) {
    console.error('Error:', err);
    if (!res.headersSent) res.status(500).send('Internal Server Error');
  } 
});

app.listen(port, () => console.log(`SECTIONS DETAILS V2 Server is running on port ${port}\n`));

File: execution/sections.py
import psycopg2
import aiohttp
from aiohttp import ClientSession, ClientTimeout, ClientError
import asyncio
import json 
from datetime import datetime, date
import string 
import random 
import psycopg2.extras

# Database Connection Variables
DB_CONFIG = {
    "host": "database-1-instance-1.cwysyy38ldpw.us-east-2.rds.amazonaws.com",
    "database": "tryAgain",
    "user": "isdsar",
    "password": "Outeclanisdsar123*",
    "port": "5432"
}

server_address = 'http://18.222.178.85'

API_URL = f'{server_address}:3019/production-isolatedsections'  # Updated API URL
API_KEY = "cf23df2207d99a74fbe169e3eba035e633b65d94"

def connect_db():
    conn = psycopg2.connect(**DB_CONFIG)
    cursor = conn.cursor()
    query = """
        SELECT pe.event_code, pe.link_id
        FROM production_evenuemap as pe
        JOIN tmevents tm ON pe.event_code = tm.event_code
        WHERE tm.on_skybox = True AND tm.temp_on_skybox = True AND tm.start_date >= current_date
        LIMIT 10; 
    """
    cursor.execute(query)
    urls = [{"event_code": row[0], "link_id": row[1]} for row in cursor.fetchall()]
    cursor.close()
    conn.close()
    return urls

async def fetch(session, params, retries=8):
    headers = {"X-API-Key": API_KEY}
    for i in range(retries + 1):
        async with session.get(API_URL, params=params, headers=headers) as response:
            # Invalid Format error 
            if response.content_type == 'application/json':
                try:
                    content = await response.json()
                except json.JSONDecodeError:
                    return f"Error: Invalid JSON received"
            else: # this is a bit unecessary 
                content = await response.text()
            #successful response 
            if response.status == 200 and 'error' not in content and '403' not in content:
                return content
            
            elif '403' in content or ('error' in content and content['error'] == 'Forbidden'):
                print(f"403: Retrying {i+1}... ")
                if i == retries:
                    return content
            else:
                if content['message'] == 'The server did not respond to the request':
                    print(f"ECONERESSET: Retrying {i+1}... ")
                    if i == retries: 
                        return content 
                else: 
                    return content # f"Error: {content}"

async def main():
    all_data = [] 
    error_data = 0  
    data_unaivalable_data = 0 
    other_error = 0 

    urls = connect_db()
    if not urls:
        return
    
    connector = aiohttp.TCPConnector(limit=len(urls))  # Set the limit to the number of URLs.
    timeout = ClientTimeout(total=15)
    async with aiohttp.ClientSession(connector=connector) as session:
        tasks = [fetch(session, url) for url in urls]
        responses = await asyncio.gather(*tasks, return_exceptions=True)

        for response, url in zip(responses, urls):
            # if isinstance(response, Exception):
            try: 
                if response['error'] == 'No response received from server': 
                    print(f"ECONERESSET: Error fetching {url}: \n{response}\n")
                    other_error += 1 
                elif response['error'] == 'Request Setup Error': 
                    print(f"BAD SET-UP ERROR: Error fetching {url}: \n{response}\n")
                    other_error += 1 
                else: 
                    print(f"UKNOWN ERROR: Error fetching {url}: \n{response}\n")
                    other_error += 1 
            except:
                if response['status'] == 403: 
                    error_data += 1 
                elif response['status'] == 200: 
                    print(f"Response from {url}: {response}")
                    all_data.append(response)
                else: 
                    other_error += 1 
    print('\n')
    print("-" * 50) 
    print('Phase 1: Successfully added data to tmEvents!')
    print('Input data size: '+str(len(urls)))
    print('Successful data size: '+str(len(all_data)))
    print('403 error size: '+str(error_data))
    print('Other error size: '+str(other_error))
    print('Not on sale yet size: '+str(data_unaivalable_data))
    print("-" * 50) 

asyncio.run(main())
